### PR TITLE
Implementing Matrix Norm for torch.norm

### PR DESCRIPTION
**Pull request by yya007-zz**
Currently, norm function only supports vector norm. This PR extends vector norm to matrix norm.

### CODE REVIEW

**Comment by yya007** [comment]
This PR is applying the comments in the # 10722 and redesign the norm function. The internal reviewer could check T28601904 for details.

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 1]
```diff
@@ -347,5 +347,45 @@ Tensor& matmul_out(Tensor &result, const Tensor & tensor1, const Tensor & tensor
...
 }
 
+Tensor frobenius_norm(const Tensor& self) {
+  IntList dim = {0, 1};
+  return at::native::frobenius_norm(self, dim);
+}
+
+Tensor frobenius_norm(const Tensor& self, IntList dim, bool keepdim) {
+  Tensor result = self.type().tensor();
+  return at::native::frobenius_norm_out(result, self, dim, keepdim);
```
Comment (SsnL) [inline aten/src/ATen/native/LinearAlgebra.cpp]: You are breaking autograd here. _out functions are not supported for autograd. You can try to back prop via frobenius_norm. it should break.

Comment (yya007) [reply aten/src/ATen/native/LinearAlgebra.cpp]: It seems that current torch.norm seems used a similar way to implement. I am wondering why that works and how we can solve this issue.

See pytorch/aten/src/ATen/native/ReduceOps.cpp line 587
Tensor norm(const Tensor& self, Scalar p, int64_t dim, bool keepdim) {
  Tensor result = self.type().tensor();
  return at::native::norm_out(result, self, p, dim, keepdim);
}

Comment (SsnL) [reply aten/src/ATen/native/LinearAlgebra.cpp]: Because norm has a custom backward defined in derivatives.yaml, but here you are relying on autograd to run backward for you.

**Comment by ezyang** [comment]
CC @erikbrinkman

**Comment by ezyang** [comment]
I'm a bit uncertain if this is the way we want to support parameter-based choice of norm. @SsnL, you've thought about this before; should we start supporting strings for selecting this sort of thing? Should we block this until enums work?

**Comment by SsnL** [comment]
@ezyang Yes enum support would be ideal. However, let's not block this bootcamp task for that. :)

File: torch/functional.py [hunk 2]
```diff
@@ -543,3 +544,241 @@ def argsort(input, dim=None, descending=False):
...
+        tensor([ 3.7417, 11.2250])
+        >>> torch.norm(d[0, :, :]), torch.norm(d[1, :, :])
+        (tensor(3.7417), tensor(11.2250))
+    """
+    ndim = input.dim()
+
+    # catch default case
+    if dim is None and out is None:
+        if p == "fro":
+            return torch._C._VariableFunctions.norm(input, 2)
```
Comment (SsnL) [inline torch/functional.py]: Hmm this is calling the wrong function, right?

File: torch/functional.py [hunk 3]
```diff
@@ -543,3 +544,241 @@ def argsort(input, dim=None, descending=False):
...
+        if isinstance(p, int):
+            return torch._C._VariableFunctions.norm(input, p)
+        if isinstance(p, float) and p != inf and p != -inf:
+            return torch._C._VariableFunctions.norm(input, p)
```
Comment (SsnL) [inline torch/functional.py]: I think this is a bit too complex. My original idea is that most of the code should live in ATen. including the arg checking code below, with only minimal Python code here doing the actual dispatch. E.g., the entire Python function body may just look like 
```py
if p == 'nuc':
  return torch._C._VariableFunctions.nuc_norm(...)
elif p == 'fro'
  return torch._C._VariableFunctions.fro_norm(...)
else:
  return torch._C._VariableFunctions.norm(...)
```

**Review by ssnl — COMMENTED** [review]
Can you also add relevant test to `test_autograd` to make sure that they are differentiable?

File: torch/functional.py (hunk 2 above)
Comment (yya007) [reply torch/functional.py]: It is intended. This follows the default case catch written in Numpy.norm. It also let current function works as the original torch.norm() considering backward compatibility.

Comment (SsnL) [reply torch/functional.py]: Sorry but I still don't quite understand here. Could you explain a bit more on why it is written this way on how it preserves BC?

File: torch/functional.py (hunk 3 above)
Comment (erikbrinkman) [reply torch/functional.py]: @ezyang when it comes to computing arbitrary norms (or distances) it seems like the appropriate c++ style is just to have different names for the functions i.e. how this PR is structured. Given that absent a reason for a better api torch should probably follow numpy's i.e. lgtm!

I will second @SsnL that python should probably just convert strings to appropriate c++ calls, maybe with some extra checking for places where python doesn't enforce the same typing that c++ does.

File: torch/functional.py (hunk 2 above)
Comment (yya007) [reply torch/functional.py]: On production, we have a function "torch.norm(input, p=2) → Tensor". Therefore, there are numerous places in the codebase where it called .norm() directly. "fro" is the default value for attr:p now as Numpy did and "fro" is equal 2-norm when calling vector norm as I mentioned in the intro. With this condition check, .norm() will have the same behavior as before.

File: torch/functional.py [hunk 4]
```diff
@@ -543,3 +544,241 @@ def argsort(input, dim=None, descending=False):
...
+            inf    max(sum(abs(x), dim=1))       max(abs(x))
+            -inf   min(sum(abs(x), dim=1))       min(abs(x))
+            0      --                            sum(x != 0)
+            1      max(sum(abs(x), dim=0))       as below
+            -1     min(sum(abs(x), dim=0))       as below
+            2      largest singular value        as below
+            -2     smallest singular value       as below
+            other  as vec norm when dim is None  sum(abs(x)**ord)**(1./ord)
+            =====  ============================  ==========================
+        dim ({int, 2-tuple of ints, 2-list of ints}): If it is a int, vector norm
```
Comment (SsnL) [inline torch/functional.py]: What happens if it is `None`?

File: torch/functional.py (hunk 3 above)
Comment (SsnL) [reply torch/functional.py]: @erikbrinkman That is the current way. However, in the long term goal, we want to have enum support in native function, and to allow specifying how to map python objects to enum values. That will make this, nn.functional.pad, and many other things much easier to write!

File: torch/functional.py (hunk 2 above)
Comment (SsnL) [reply torch/functional.py]: I see. But ideally, if you call `torch._C._VariableFunctions.frobenius_norm` with `dim` being `None`, it should do the same thing as 2-norm in c++, right?

**Comment by yya007** [comment]
@SsnL Autograd test has been added.

File: torch/functional.py [hunk 5]
```diff
@@ -599,3 +600,111 @@ def argsort(input, dim=None, descending=False):
...
+            The following norms can be calculated:
+            =====  ============================  ==========================
+            ord    matrix norm                   vector norm
+            =====  ============================  ==========================
+            None   Frobenius norm                2-norm
+            'fro'  Frobenius norm                --
+            'nuc'  nuclear norm                  --
+            other  as vec norm when dim is None  sum(abs(x)**ord)**(1./ord)
+            =====  ============================  ==========================
+        dim ({int, 2-tuple of ints, 2-list of ints}, optional): If it is a int,
```
Comment (li-roy) [inline torch/functional.py]: nit: an int

File: torch/functional.py [hunk 6]
```diff
@@ -599,3 +600,111 @@ def argsort(input, dim=None, descending=False):
...
+
+    # catch default case
+    if dim is None and out is None:
+        if isinstance(p, int) or (isinstance(p, float) and p != inf and p != -inf):
+            return torch._C._VariableFunctions.norm(input, p)
+        if p == "fro":
+            return torch._C._VariableFunctions.frobenius_norm(input)
+
+    if dim is None:
+        dim = tuple(range(ndim))
```
Comment (SsnL) [inline torch/functional.py]: Why this?

File: torch/functional.py [hunk 7]
```diff
@@ -599,3 +600,111 @@ def argsort(input, dim=None, descending=False):
...
+        if out is None:
+            return torch._C._VariableFunctions.frobenius_norm(input, dim, keepdim=keepdim)
+        return torch._C._VariableFunctions.frobenius_norm(input, dim, keepdim=keepdim, out=out)
+    elif p == "nuc":
+        if out is None:
+            torch._C._VariableFunctions.nuclear_norm(input, keepdim=keepdim)
+        return torch._C._VariableFunctions.nuclear_norm(input, keepdim=keepdim, out=out)
+    else:
+        if out is None:
+            return torch._C._VariableFunctions.norm(input, p, dim[-1], keepdim=keepdim)
```
Comment (SsnL) [inline torch/functional.py]: You are hiding incorrect argument by using `dim[-1]`.

File: torch/functional.py [hunk 8]
```diff
@@ -599,3 +600,111 @@ def argsort(input, dim=None, descending=False):
...
+    elif isinstance(dim, list):
+        dim = tuple(dim)
+    elif not isinstance(dim, tuple):
+        try:
+            dim = int(dim)
+        except Exception:
+            raise TypeError(
+                "'dim' must be None, an integer or a tuple/list of integers"
+            )
+        dim = (dim,)
```
Comment (SsnL) [inline torch/functional.py]: Why do we need to parse `dim` ourselves? Can't we just always passing it to the C functions, except in `nuc` case we can do some manual checking because it doesn't support `dim` argument.

File: torch/functional.py [hunk 9]
```diff
@@ -599,3 +600,111 @@ def argsort(input, dim=None, descending=False):
...
+        >>> d = torch.arange(8, dtype= torch.float).reshape(2,2,2)
+        >>> torch.norm(d, dim=(1,2))
+        tensor([ 3.7417, 11.2250])
+        >>> torch.norm(d[0, :, :]), torch.norm(d[1, :, :])
+        (tensor(3.7417), tensor(11.2250))
+    """
+    ndim = input.dim()
+
+    # catch default case
+    if dim is None and out is None:
```
Comment (SsnL) [inline torch/functional.py]: If you just pass `dim` to the C function, none of these will be needed, right?

File: torch/functional.py [hunk 10]
```diff
@@ -599,3 +600,111 @@ def argsort(input, dim=None, descending=False):
...
+        calculated when the input tensor only has one dimension. If the input
+        tensor has more than two dimensions, the vector norm will be applied to
+        last dimension.
+        keepdim (bool): whether the output tensors have :attr:`dim`
+            retained or not. Ignored if ``dim=None and out=None``.
+        out (Tensor, optional) – the output tensor. Ignored if
+        ``dim=None and out=None``.
+
+    Example::
+        >>> import torch
```
Comment (li-roy) [inline torch/functional.py]: Can we get rid of some of these examples? It will won't look good on the docs page. I feel like the examples should illustrate proper usage of the function, it doesn't need to cover every code path.

**Review by li-roy — APPROVED** [review]
I feel like this is good to go.
Imo, we should accept the logic in python as an intermediate step, but since @SsnL has given more thought into this, I'll defer to you for the final decision.

File: torch/functional.py [hunk 11]
```diff
@@ -599,3 +600,111 @@ def argsort(input, dim=None, descending=False):
...
+        if isinstance(p, int) or (isinstance(p, float) and p != inf and p != -inf):
+            return torch._C._VariableFunctions.norm(input, p)
+        if p == "fro":
+            return torch._C._VariableFunctions.frobenius_norm(input)
+
+    if dim is None:
+        dim = tuple(range(ndim))
+    elif isinstance(dim, list):
+        dim = tuple(dim)
+    elif not isinstance(dim, tuple):
```
Comment (li-roy) [inline torch/functional.py]: Can we try to simplify the logic here, a lot of this seems unneeded.

**Review by li-roy — CHANGES_REQUESTED** [review]
please resolve comments

File: torch/functional.py (hunk 6 above)
Comment (li-roy) [reply torch/functional.py]: I think this is needed to support lists of integers, but I think we should probably drop list to align with the rest of the api.

Comment (li-roy) [reply torch/functional.py]: nvm sorry, I'm wrong

**Comment by yya007** [comment]
@li-roy  Thank you for the comments. I have resolved those.

**Review by li-roy — APPROVED** [review]

File: torch/functional.py [hunk 12]
```diff
@@ -599,3 +601,81 @@ def argsort(input, dim=None, descending=False):
     if dim is None:
         return torch.sort(input, -1, descending)[1]
     return torch.sort(input, dim, descending)[1]
+
+
+def norm(input, p="fro", dim=None, keepdim=False, out=None):
```
Comment (ezyang) [inline torch/functional.py]: p is a bit of an unfortunate name for the kwarg now, but I guess it'll have to do (since changing it would be BC-breaking)
